lib/Sema/CSDiagnostics.cpp

```diff
@@ -997,6 +1001,18 @@ bool MemberAccessOnOptionalBaseFailure::diagnoseAsError() {
   if (overload && overload->openedType->getOptionalObjectType())
     resultIsOptional = true;
 
+  // Tailored logic to get the base optional type when the failure is
+  // for a keypath component member.
+  if (auto componentPathElt =
+          locator->getLastElementAs<LocatorPathElt::KeyPathComponent>()) {
+    auto keyPathExpr = castToExpr<KeyPathExpr>(anchor);
+
+    auto componentType = cs.getType(keyPathExpr, componentPathElt->getIndex() - 1);
```

Comment:
We want to avoid querying the constraint system during diagnostics, as its state may not be the same as when the fix was originally recorded. Note also that this doesn't handle the case where the optional base is the root of the key path, e.g `\String?.count`.

I think the easiest solution here might be to have the `UnwrapOptionalBase` fix store the optional base type rather than trying to recover it later. Note though that we may still need some custom logic to get the right source location for where to insert the postfix `?` and `!` fix-its in the key path (and to disable the fix-its for an optional key path root). Could you adjust the tests to make sure we're inserting them in the right place?

Reply:
> Note though that we may still need some custom logic to get the right source location for where to insert the postfix ? and ! fix-its in the key path (and to disable the fix-its for an optional key path root). Could you adjust the tests to make sure we're inserting them in the right place?

Sure, I'm adjusting that right now :) 

> We want to avoid querying the constraint system during diagnostics, as its state may not be the same as when the fix was originally recorded. Note also that this doesn't handle the case where the optional base is the root of the key path, e.g \String?.count.

Right, I'll make the adjustments :) Thank you

Reply:
Also @hamishknight 

> Note also that this doesn't handle the case where the optional base is the root of the key path, e.g \String?.count

I can't see the case where that will require this fix, do you have an example?

Reply:
@LucianoPAlmeida Does `\String?.count` not create the fix? As I understand it, we should add a member constraint for `count` with a base of `String?`, requiring the member-on-optional fix.

Reply:
Ahh sorry, I just misunderstood ... because I was thinking of the fixes and how in this situation it will help to add an extra `?` or `!` ...
But yes it creates it

Reply:
Right yeah, we'll want to emit the error about the incorrect optionality, but we'll want to avoid telling the user to add `!` or `?` :)

Reply:
That's what I thought :) Thanks @hamishknight, I'll make some quick adjustments in another PR and get back to this